File: python/features/steps/comment_steps.py
"""Behave steps for issue comments."""

from __future__ import annotations

from datetime import datetime, timezone

from behave import given, then, when

from features.steps.shared import (
    build_issue,
    load_project_directory,
    read_issue_file,
    run_cli,
    write_issue_file,
)
from kanbus.issue_comment import (
    IssueCommentError,
    delete_comment,
    ensure_issue_comment_ids,
    update_comment,
)
from kanbus.models import IssueComment


@given('the current user is "dev@example.com"')
def given_current_user(context: object) -> None:
    overrides = dict(getattr(context, "environment_overrides", {}) or {})
    overrides["KANBUS_USER"] = "dev@example.com"
    context.environment_overrides = overrides


@when('I run "kanbus comment kanbus-aaa \\"First comment\\""')
def when_comment_first(context: object) -> None:
    run_cli(context, 'kanbus comment kanbus-aaa "First comment"')


@when('I run "kanbus comment kanbus-aaa \\"Second comment\\""')
def when_comment_second(context: object) -> None:
    run_cli(context, 'kanbus comment kanbus-aaa "Second comment"')


@when('I run "kanbus comment kanbus-missing \\"Missing issue note\\""')
def when_comment_missing(context: object) -> None:
    run_cli(context, 'kanbus comment kanbus-missing "Missing issue note"')


@when('I run "kanbus comment kanbus-note \\"Searchable comment\\""')
def when_comment_note(context: object) -> None:
    run_cli(context, 'kanbus comment kanbus-note "Searchable comment"')


@when('I run "kanbus comment kanbus-dup \\"Dup keyword\\""')
def when_comment_dup(context: object) -> None:
    run_cli(context, 'kanbus comment kanbus-dup "Dup keyword"')


@then('issue "kanbus-aaa" should have 1 comment')
def then_issue_has_one_comment(context: object) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, "kanbus-aaa")
    assert len(issue.comments) == 1


@then('the latest comment should have author "dev@example.com"')
def then_latest_author(context: object) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, "kanbus-aaa")
    assert issue.comments[-1].author == "dev@example.com"


@then('the latest comment should have text "First comment"')
def then_latest_text(context: object) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, "kanbus-aaa")
    assert issue.comments[-1].text == "First comment"


@then("the latest comment should have a created_at timestamp")
def then_latest_timestamp(context: object) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, "kanbus-aaa")
    assert issue.comments[-1].created_at is not None


@then(
    'issue "kanbus-aaa" should have comments in order "First comment", "Second comment"'
)
def then_comments_order(context: object) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, "kanbus-aaa")
    texts = [comment.text for comment in issue.comments]
    assert texts == ["First comment", "Second comment"]


@given('an issue "{identifier}" exists with a comment missing an id')
def given_issue_with_missing_comment_id(context: object, identifier: str) -> None:
    project_dir = load_project_directory(context)
    issue = build_issue(identifier, "Legacy comments", "task", "open", None, [])
    comment = IssueComment(
        id=None,
        author="dev@example.com",
        text="Legacy note",
        created_at=datetime(2026, 2, 11, tzinfo=timezone.utc),
    )
    issue = issue.model_copy(update={"comments": [comment]})
    write_issue_file(project_dir, issue)


@given(
    'an issue "{identifier}" exists with comment id "{comment_id}" and text "{text}"'
)
def given_issue_with_comment_id(
    context: object, identifier: str, comment_id: str, text: str
) -> None:
    project_dir = load_project_directory(context)
    issue = build_issue(identifier, "Commented issue", "task", "open", None, [])
    comment = IssueComment(
        id=comment_id,
        author="dev@example.com",
        text=text,
        created_at=datetime(2026, 2, 11, tzinfo=timezone.utc),
    )
    issue = issue.model_copy(update={"comments": [comment]})
    write_issue_file(project_dir, issue)


@given('an issue "{identifier}" exists with comment ids "{first_id}" and "{second_id}"')
def given_issue_with_comment_ids(
    context: object, identifier: str, first_id: str, second_id: str
) -> None:
    project_dir = load_project_directory(context)
    issue = build_issue(identifier, "Ambiguous comments", "task", "open", None, [])
    comments = [
        IssueComment(
            id=first_id,
            author="dev@example.com",
            text="First",
            created_at=datetime(2026, 2, 11, tzinfo=timezone.utc),
        ),
        IssueComment(
            id=second_id,
            author="dev@example.com",
            text="Second",
            created_at=datetime(2026, 2, 11, tzinfo=timezone.utc),
        ),
    ]
    issue = issue.model_copy(update={"comments": comments})
    write_issue_file(project_dir, issue)


@when('I ensure comment ids for "{identifier}"')
def when_ensure_comment_ids(context: object, identifier: str) -> None:
    project_dir = load_project_directory(context)
    context.updated_issue = ensure_issue_comment_ids(project_dir, identifier)


@when('I update comment "{comment_id}" on "{identifier}" to "{text}"')
def when_update_comment(
    context: object, comment_id: str, identifier: str, text: str
) -> None:
    project_dir = load_project_directory(context)
    context.updated_issue = update_comment(project_dir, identifier, comment_id, text)


@when('I delete comment "{comment_id}" on "{identifier}"')
def when_delete_comment(context: object, comment_id: str, identifier: str) -> None:
    project_dir = load_project_directory(context)
    context.updated_issue = delete_comment(project_dir, identifier, comment_id)


@when('I attempt to update comment "{comment_id}" on "{identifier}" to "{text}"')
def when_attempt_update_comment(
    context: object, comment_id: str, identifier: str, text: str
) -> None:
    project_dir = load_project_directory(context)
    if comment_id == "<empty>":
        comment_id = ""
    try:
        context.updated_issue = update_comment(
            project_dir, identifier, comment_id, text
        )
        context.last_error = None
    except IssueCommentError as error:
        context.last_error = error


@when('I attempt to delete comment "{comment_id}" on "{identifier}"')
def when_attempt_delete_comment(
    context: object, comment_id: str, identifier: str
) -> None:
    project_dir = load_project_directory(context)
    if comment_id == "<empty>":
        comment_id = ""
    try:
        context.updated_issue = delete_comment(project_dir, identifier, comment_id)
        context.last_error = None
    except IssueCommentError as error:
        context.last_error = error


@then('issue "{identifier}" should have comment ids assigned')
def then_issue_comment_ids_assigned(context: object, identifier: str) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.comments, "expected at least one comment"
    assert all(comment.id for comment in issue.comments)


@then('issue "{identifier}" should have comment text "{text}"')
def then_issue_comment_text(context: object, identifier: str, text: str) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert issue.comments, "expected at least one comment"
    assert issue.comments[0].text == text


@then('issue "{identifier}" should have {count:d} comments')
def then_issue_comment_count(context: object, identifier: str, count: int) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert len(issue.comments) == count


@then('the last comment operation should fail with "{message}"')
def then_comment_operation_failed(context: object, message: str) -> None:
    error = getattr(context, "last_error", None)
    assert error is not None, "expected an error"
    assert message in str(error)

File: rust/features/steps/comment_steps.rs
use std::fs;
use std::path::PathBuf;

use chrono::Utc;
use cucumber::{given, then, when};

use kanbus::cli::run_from_args_with_output;
use kanbus::file_io::load_project_directory;
use kanbus::models::{IssueComment, IssueData};

use crate::step_definitions::initialization_steps::KanbusWorld;

fn load_project_dir(world: &KanbusWorld) -> PathBuf {
    let cwd = world.working_directory.as_ref().expect("cwd");
    load_project_directory(cwd).expect("project dir")
}

fn load_issue(project_dir: &PathBuf, identifier: &str) -> IssueData {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    let contents = fs::read_to_string(issue_path).expect("read issue");
    serde_json::from_str(&contents).expect("parse issue")
}

fn save_issue(project_dir: &PathBuf, issue: &IssueData) {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{}.json", issue.identifier));
    let contents = serde_json::to_string_pretty(issue).expect("serialize issue");
    fs::write(issue_path, contents).expect("write issue");
}

#[given("the current user is \"dev@example.com\"")]
fn given_current_user(_world: &mut KanbusWorld) {
    std::env::set_var("KANBUS_USER", "dev@example.com");
}

#[then("issue \"kanbus-aaa\" should have 1 comment")]
fn then_issue_has_one_comment(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issue = load_issue(&project_dir, "kanbus-aaa");
    assert_eq!(issue.comments.len(), 1);
}

#[then("the latest comment should have author \"dev@example.com\"")]
fn then_latest_author(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issue = load_issue(&project_dir, "kanbus-aaa");
    let latest = issue.comments.last().expect("comment");
    assert_eq!(latest.author, "dev@example.com");
}

#[then("the latest comment should have text \"First comment\"")]
fn then_latest_text(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issue = load_issue(&project_dir, "kanbus-aaa");
    let latest = issue.comments.last().expect("comment");
    assert_eq!(latest.text, "First comment");
}

#[then("the latest comment should have a created_at timestamp")]
fn then_latest_timestamp(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issue = load_issue(&project_dir, "kanbus-aaa");
    let latest = issue.comments.last().expect("comment");
    assert!(latest.created_at.timestamp() > 0);
}

#[then("issue \"kanbus-aaa\" should have comments in order \"First comment\", \"Second comment\"")]
fn then_comments_order(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issue = load_issue(&project_dir, "kanbus-aaa");
    let texts: Vec<String> = issue
        .comments
        .iter()
        .map(|comment| comment.text.clone())
        .collect();
    assert_eq!(
        texts,
        vec!["First comment".to_string(), "Second comment".to_string()]
    );
}

// Additional Given steps for comment manipulation tests

#[given(expr = "an issue {string} exists with a comment missing an id")]
fn given_issue_with_comment_missing_id(world: &mut KanbusWorld, identifier: String) {
    let project_dir = load_project_dir(world);
    let timestamp = Utc::now();
    let issue = IssueData {
        identifier: identifier.clone(),
        title: "Test Issue".to_string(),
        description: String::new(),
        issue_type: "task".to_string(),
        status: "open".to_string(),
        priority: 2,
        assignee: None,
        creator: None,
        parent: None,
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: vec![IssueComment {
            id: None, // missing id
            author: "user@example.com".to_string(),
            text: "Legacy comment".to_string(),
            created_at: timestamp,
        }],
        created_at: timestamp,
        updated_at: timestamp,
        closed_at: None,
        custom: std::collections::BTreeMap::new(),
    };
    save_issue(&project_dir, &issue);
}

#[given(expr = "an issue {string} exists with comment id {string} and text {string}")]
fn given_issue_with_comment_id_and_text(
    world: &mut KanbusWorld,
    identifier: String,
    comment_id: String,
    text: String,
) {
    let project_dir = load_project_dir(world);
    let timestamp = Utc::now();
    let issue = IssueData {
        identifier: identifier.clone(),
        title: "Test Issue".to_string(),
        description: String::new(),
        issue_type: "task".to_string(),
        status: "open".to_string(),
        priority: 2,
        assignee: None,
        creator: None,
        parent: None,
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: vec![IssueComment {
            id: Some(comment_id),
            author: "user@example.com".to_string(),
            text,
            created_at: timestamp,
        }],
        created_at: timestamp,
        updated_at: timestamp,
        closed_at: None,
        custom: std::collections::BTreeMap::new(),
    };
    save_issue(&project_dir, &issue);
}

#[given(expr = "an issue {string} exists with comment ids {string} and {string}")]
fn given_issue_with_two_comment_ids(
    world: &mut KanbusWorld,
    identifier: String,
    id1: String,
    id2: String,
) {
    let project_dir = load_project_dir(world);
    let timestamp = Utc::now();
    let issue = IssueData {
        identifier: identifier.clone(),
        title: "Test Issue".to_string(),
        description: String::new(),
        issue_type: "task".to_string(),
        status: "open".to_string(),
        priority: 2,
        assignee: None,
        creator: None,
        parent: None,
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: vec![
            IssueComment {
                id: Some(id1),
                author: "user@example.com".to_string(),
                text: "First".to_string(),
                created_at: timestamp,
            },
            IssueComment {
                id: Some(id2),
                author: "user@example.com".to_string(),
                text: "Second".to_string(),
                created_at: timestamp,
            },
        ],
        created_at: timestamp,
        updated_at: timestamp,
        closed_at: None,
        custom: std::collections::BTreeMap::new(),
    };
    save_issue(&project_dir, &issue);
}

// When steps for comment operations

#[when(expr = "I ensure comment ids for {string}")]
fn when_ensure_comment_ids(world: &mut KanbusWorld, identifier: String) {
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let argv = vec![
        "kanbus".to_string(),
        "comment".to_string(),
        "ensure-ids".to_string(),
        identifier,
    ];

    match run_from_args_with_output(argv, cwd.as_path()) {
        Ok(output) => {
            world.exit_code = Some(0);
            world.stdout = Some(output.stdout);
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

#[when(expr = "I update comment {string} on {string} to {string}")]
fn when_update_comment(
    world: &mut KanbusWorld,
    comment_prefix: String,
    identifier: String,
    new_text: String,
) {
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let argv = vec![
        "kanbus".to_string(),
        "comment".to_string(),
        "update".to_string(),
        identifier,
        comment_prefix,
        new_text,
    ];

    match run_from_args_with_output(argv, cwd.as_path()) {
        Ok(output) => {
            world.exit_code = Some(0);
            world.stdout = Some(output.stdout);
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

#[when(expr = "I delete comment {string} on {string}")]
fn when_delete_comment(world: &mut KanbusWorld, comment_prefix: String, identifier: String) {
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let argv = vec![
        "kanbus".to_string(),
        "comment".to_string(),
        "delete".to_string(),
        identifier,
        comment_prefix,
    ];

    match run_from_args_with_output(argv, cwd.as_path()) {
        Ok(output) => {
            world.exit_code = Some(0);
            world.stdout = Some(output.stdout);
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

#[when(expr = "I attempt to update comment {string} on {string} to {string}")]
fn when_attempt_update_comment(
    world: &mut KanbusWorld,
    comment_prefix: String,
    identifier: String,
    new_text: String,
) {
    // Replace <empty> with empty string
    let prefix = if comment_prefix == "<empty>" {
        String::new()
    } else {
        comment_prefix
    };
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let argv = vec![
        "kanbus".to_string(),
        "comment".to_string(),
        "update".to_string(),
        identifier,
        prefix,
        new_text,
    ];

    match run_from_args_with_output(argv, cwd.as_path()) {
        Ok(output) => {
            world.exit_code = Some(0);
            world.stdout = Some(output.stdout);
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

#[when(expr = "I attempt to delete comment {string} on {string}")]
fn when_attempt_delete_comment(
    world: &mut KanbusWorld,
    comment_prefix: String,
    identifier: String,
) {
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let argv = vec![
        "kanbus".to_string(),
        "comment".to_string(),
        "delete".to_string(),
        identifier,
        comment_prefix,
    ];

    match run_from_args_with_output(argv, cwd.as_path()) {
        Ok(output) => {
            world.exit_code = Some(0);
            world.stdout = Some(output.stdout);
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

#[when(expr = "I attempt to ensure comment ids for {string}")]
fn when_attempt_ensure_comment_ids(world: &mut KanbusWorld, identifier: String) {
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let argv = vec![
        "kanbus".to_string(),
        "comment".to_string(),
        "ensure-ids".to_string(),
        identifier,
    ];

    match run_from_args_with_output(argv, cwd.as_path()) {
        Ok(output) => {
            world.exit_code = Some(0);
            world.stdout = Some(output.stdout);
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

// Then steps for comment assertions

#[then(expr = "issue {string} should have comment ids assigned")]
fn then_issue_has_comment_ids(world: &mut KanbusWorld, identifier: String) {
    let project_dir = load_project_dir(world);
    let issue = load_issue(&project_dir, &identifier);
    for comment in &issue.comments {
        assert!(comment.id.is_some(), "Comment missing id");
    }
}

#[then(expr = "issue {string} should have comment text {string}")]
fn then_issue_has_comment_text(world: &mut KanbusWorld, identifier: String, expected_text: String) {
    let project_dir = load_project_dir(world);
    let issue = load_issue(&project_dir, &identifier);
    assert_eq!(issue.comments.len(), 1);
    assert_eq!(issue.comments[0].text, expected_text);
}

#[then(expr = "issue {string} should have {int} comments")]
fn then_issue_has_comment_count(world: &mut KanbusWorld, identifier: String, count: usize) {
    let project_dir = load_project_dir(world);
    let issue = load_issue(&project_dir, &identifier);
    assert_eq!(issue.comments.len(), count);
}

#[then(expr = "the last comment operation should fail with {string}")]
fn then_last_comment_op_fails(world: &mut KanbusWorld, expected_error: String) {
    assert_eq!(world.exit_code, Some(1), "Expected operation to fail");
    let stderr = world.stderr.as_ref().expect("stderr");
    assert!(
        stderr.contains(&expected_error),
        "Expected error containing '{}', got: {}",
        expected_error,
        stderr
    );
}
